### PR TITLE
Fix: node storage literal ids assignment

### PR DESCRIPTION
**Pull request by liss-h**
Additional PR to complete #146

### CODE REVIEW

**Comment by liss-h** [comment]
Don't mind the massive amount of commits, I branched from the other PR

File: src/rdf4cpp/rdf/storage/node/reference_node_storage/ReferenceNodeStorageBackend.cpp [hunk 1]
```diff
@@ -155,42 +154,71 @@ static identifier::NodeID lookup_or_insert_impl(typename Storage::BackendView co
...
     return view.visit(
             [this](view::LexicalFormLiteralBackendView const &lexical) noexcept {
-                auto const next_id_func = [this, datatype = lexical.datatype_id]() noexcept {
-                    return identifier::NodeID{next_literal_id++, identifier::iri_node_id_to_literal_type(datatype)};
+                auto const datatype = identifier::iri_node_id_to_literal_type(lexical.datatype_id);
+                assert(!this->has_specialized_storage_for(datatype));
+
+                auto const next_id_func = [this, datatype]() noexcept {
+                    auto const id = next_fallback_literal_id_.fetch_add(1, std::memory_order_relaxed);
+                    if (id >= 1ul << LiteralID::width) [[unlikely]] {
```
Comment (bigerl) [inline src/rdf4cpp/rdf/storage/node/reference_node_storage/ReferenceNodeStorageBackend.cpp]:
```suggestion
                    if (id >= (1ul << LiteralID::width)) [[unlikely]] {
```

for readability.

File: src/rdf4cpp/rdf/storage/node/reference_node_storage/ReferenceNodeStorageBackend.cpp [hunk 2]
```diff
@@ -155,42 +154,71 @@ static identifier::NodeID lookup_or_insert_impl(typename Storage::BackendView co
...
                 return lookup_or_insert_impl<true>(lexical, fallback_literal_storage_, next_id_func);
             },
             [this](view::ValueLiteralBackendView const &any) noexcept {
+                assert(this->has_specialized_storage_for(any.datatype));
+
                 auto const next_id_func = [this, datatype = any.datatype]() noexcept {
-                    return identifier::NodeID{next_literal_id++, datatype};
+                    return visit_specialized(next_specialized_literal_ids_, datatype, [datatype](auto &lit_id) {
+                        auto const id = lit_id.fetch_add(1, std::memory_order_relaxed);
+                        if (id >= 1ul << LiteralID::width) [[unlikely]] {
```
Comment (bigerl) [inline src/rdf4cpp/rdf/storage/node/reference_node_storage/ReferenceNodeStorageBackend.cpp]: see above

File: src/rdf4cpp/rdf/storage/node/reference_node_storage/ReferenceNodeStorageBackend.cpp [hunk 3]
```diff
@@ -155,42 +154,71 @@ static identifier::NodeID lookup_or_insert_impl(typename Storage::BackendView co
...
 }
 
 identifier::NodeID ReferenceNodeStorageBackend::find_or_make_id(view::BNodeBackendView const &view) noexcept {
-    return lookup_or_insert_impl<true>(view, bnode_storage_,
-                                       [this]() noexcept {
-                                           return next_bnode_id++;
-                                       });
+    return lookup_or_insert_impl<true>(view, bnode_storage_, [this]() noexcept {
+        auto const id = next_bnode_id_.fetch_add(1, std::memory_order_relaxed);
+        if (id >= 1ul << NodeID::width) [[unlikely]] {
```
Comment (bigerl) [inline src/rdf4cpp/rdf/storage/node/reference_node_storage/ReferenceNodeStorageBackend.cpp]: see above

File: src/rdf4cpp/rdf/storage/node/reference_node_storage/ReferenceNodeStorageBackend.cpp [hunk 4]
```diff
@@ -155,42 +154,71 @@ static identifier::NodeID lookup_or_insert_impl(typename Storage::BackendView co
...
 }
 
 identifier::NodeID ReferenceNodeStorageBackend::find_or_make_id(view::VariableBackendView const &view) noexcept {
-    return lookup_or_insert_impl<true>(view, variable_storage_,
-                                       [this]() noexcept {
-                                           return next_variable_id++;
-                                       });
+    return lookup_or_insert_impl<true>(view, variable_storage_, [this]() noexcept {
+        auto const id = next_variable_id_.fetch_add(1, std::memory_order_relaxed);
+        if (id >= 1ul << NodeID::width) [[unlikely]] {
```
Comment (bigerl) [inline src/rdf4cpp/rdf/storage/node/reference_node_storage/ReferenceNodeStorageBackend.cpp]: see above

**Review by bigerl — CHANGES_REQUESTED** [review]
Thx for the PR. The patch looks good. It's a bit unfortunate that you had to replace the access by type with an access by index. Still, I think it is better to use the array here because using a tuple instead would probably prolong compilation again. 

The proposed change is only of cosmetic nature. Otherwise, the PR can be merged.

**Review by bigerl — APPROVED** [review]

**Pull request merged**